### PR TITLE
feat: Introduce a "bordered" variant for all text-like fields

### PR DESCRIPTION
**Pull request by gnapse**
## Short description

This would allow use the Reactist component for the views that use this different style of text fields (e.g. Todoist and Twist login/onboarding pages).

- [Related discussion in Twist](https://twist.com/a/1585/ch/190200/t/3456577/c/76152665)

### A note on testing

This functionality is almost purely style-related[^1], it is a bit pointless to test in jest. I could test that the correct classes are being added here and there, but that provides very little confidence, since so much could be broken by changes in the style rules applied to those classes.

Given my upcoming plans to work in visual regression testing, I'm putting my bets on that for really testing this in a much better and confident way. But if you think there's something we could do that would add value, let me know.

[^1]: Even some interaction elements that we'd like to make sure we test somehow (e.g. clicking inside blank parts inside bordered area focuses on the field) is achieved via making sure with CSS that the label expands to those blank areas, and "clicking in the label to focus" is a functionality that browsers give us for free, so there's no JS involved in making that work.

### A note on some implementation details

The way `BaseField` works is proving more and more challenging to make it work with all the smaller details and nuances of each of the components that relies on it. This may need to be revisited to see how we can better provide a framework for field-like components, while being more flexible in reusing styles and allowing each component to customize things to its needs even more.

Given how prone this is to breaking if we do it now or soon, I am also putting my bets on visual regression testing. If we manage to have it place, it would allow us to make bigger internal refactors with more confidence that we will not break things visually.

## Test plan for reviewers

Some things to pay attention to while reviewing this:

1. In all 3 types of fields (`TextField`, `TextArea`, `PasswordField`)
    - [x] Make sure the border is red when marked with an error message (regardless of the border being outside with the new variant, or the border being only around the typing area, as it has always been before this PR)
    - [x] Similar to the point above, but taking a look at how the border color changes slightly when focused vs. not focused. Regardless of where the border is (containing the labels vs. not containing the labels, the border should react to focus/blur just the same)
    - [x] When using the new bordered mode, see that clicking on the blank area to the right of the label focuses the field.
2. For `PasswordField`
    - [x] Pay some attention to how the "toggle password visibility" button looks like in bordered vs regular style modes. Does it all look good?
3. For `TextArea`
    - [ ] Pay some attention to how the text area allows to be resized, but only vertically. I want your opinion on this. It was not good to allow to resize it horizontally while in bordered mode (because the wrapper around it would not adjust accordingly). Should we allow resizing fully when in the default (non-bordered) mode?

## PR Checklist

-   [ ] Added tests for bugs / new features
-   [x] Updated docs (storybooks, readme)
-   [x] Executed `npm run validate` and made sure no errors / warnings were shown
-   [x] Described changes in `CHANGELOG.md`
-   [ ] Bumped version in `package.json` and `package-lock.json` (`npm --no-git-tag-version version <major|minor|patch>`) [ref](https://docs.npmjs.com/cli/v6/commands/npm-version)
-   [ ] Updated all static build artifacts (`npm run build-all`)

## Versioning

New minor release

## Demo

<ul>

<li>
<details>
<summary><code>TextField</code> 🔻</summary>

<table><tr><td>

CleanShot 2022-06-02 at 16 29 28

</td><td>

CleanShot 2022-06-02 at 16 26 32@2x

</td></tr></table>

</details>
</li>

<li>
<details>
<summary><code>PasswordField</code> 🔻</summary>

<table><tr><td>

CleanShot 2022-06-02 at 16 43 33

</td><td>

CleanShot 2022-06-02 at 16 44 11@2x

</td></tr></table>

</details>
</li>

<li>
<details>
<summary><code>TextArea</code> 🔻</summary>

<table><tr><td>

CleanShot 2022-06-02 at 16 47 37

</td><td>

CleanShot 2022-06-02 at 16 48 43@2x

</td></tr></table>

</details>
</li>

</ul>

### CODE REVIEW

**Comment by frankieyan** [comment]
Hey @gnapse, if it's not too much of a rush I'll review this Monday as I'll be shutting down my week early. I don't think I can reroll with the frontend-product group though as Paul is also assigned.

**Comment by gnapse** [comment]
> Hey @gnapse, if it's not too much of a rush I'll review this Monday as I'll be shutting down my week early. I don't think I can reroll with the frontend-product group though as Paul is also assigned.

Oh for sure. There's absolutely no rush on this. I actually want to take time, like I said, to run this in context in Twist or Todoist's login screens locally, before merging. This is a slow one. Take your time.

**Comment by pauloslund** [comment]
Sorry for the delay on this! I haven't forgotten about it, and it's on my list to look at tomorrow

**Review by frankieyan — COMMENTED** [review]
Tested this in storybook and it's looking pretty good, @gnapse 

>  It was not good to allow to resize it horizontally while in bordered mode (because the wrapper around it would not adjust accordingly). Should we allow resizing fully when in the default (non-bordered) mode?

What if we set this based on the max-width the user passes in? We could then fall back to something like `min-content` and have the container constrained to the size of the textarea, allowing it to be resized horizontally:

https://user-images.githubusercontent.com/8531248/172570558-ff68a7b5-3da8-41c4-977f-58f434aa0093.mov

**Review by pauloslund — COMMENTED** [review]
I don't really have many comments on the code itself, it feels well thought out and I appreciate the comments!

> Pay some attention to how the text area allows to be resized, but only vertically.

As someone who likes having a lot of space to write, I'm not a huge fan of this restriction. I think Frankie's suggestion about it being dependant on maxWidth could be a good compromise.

> The way `BaseField` works is proving more and more challenging to make it work with all the smaller details and nuances of each of the components that relies on it. This may need to be revisited to see how we can better provide a framework for field-like components, while being more flexible in reusing styles and allowing each component to customize things to its needs even more.

Long term, I think refactoring the interface for using `BaseField` to composition instead of props would likely be the best approach for ensuring we have the flexibility we're looking for.

This is a simplified example of what I'm thinking it would look like in use:
```jsx
function TextField({ label, secondaryLabel, auxiliaryLabel, ...props }) {
  return (
    <BaseField.Container>
      <BaseField.LabelContainer>
        <BaseField.Label>{label}</BaseField.Label>
        {secondaryLabel ? (
          <BaseField.Label variant="secondary">
            &nbsp;({secondaryLabel})
          </BaseField.Label>
        ) : null}
      </BaseField.LabelContainer>

      {auxiliaryLabel ? (
        <BaseField.Label variant="auxiliary"></BaseField.Label>
      ) : null}

      <BaseField.Input {...props} />
      {/* message, hint, etc */}
    </BaseField.Container>
  );
}
```

I think you brought up something similar in [this thread](https://twist.com/a/1585/ch/190200/t/3490506/), however the difference would be that we still export the props driven `TextField`, `TextArea`, and `PasswordField` components. This way  we could still enforce design requirements, while not requiring a large amount of boilerplate for implementing forms[^1].

```jsx
function LoginForm({ email, password, handleChangeEmail, handleChangePassword, handleSubmit }) {
  return (
    <form onSubmit={handleSubmit}>
      <TextField value={email} onChange={handleChangeEmail} label="Email" />
      <PasswordField value={password} onChange={handleChangePassword} label="Password" />
    </form>
  )
}
```


[^1]: Personally, I would still advocate for exporting the `BaseField` components to consumers. I believe optimizing for change by increasing flexibility is beneficial, but that's potentially a larger/different discussion.

File: src/new-components/base-field/base-field.module.css
Comment (gnapse) [reply src/new-components/base-field/base-field.module.css:18]: Not sure what I was thinking. Indeed, this can be simplified, and there's no need for this `.text` class name. I'll do it.

**Comment by gnapse** [comment]
@frankieyan @pauloslund I addressed your concerns:

- Full ability to resize of `TextArea` horizontally and not just vertically (within the `maxWidth` constraints).
- Clicking in the gap between label and field now makes the field keep focus.
- Remove some unneeded complexity around the styles (no need for the `.text` class name, point out by Paul)

Can you both take a look again? I'd appreciate having the entire test plan checked off, or any new observations that you may have, before approving this, if any.

**Review by frankieyan — APPROVED** [review]
Looks good @gnapse! I didn't check off the last point in the test plan as it's still talking about the previous restrictions, but I can no longer find the gap in-between, and resizing also feels good 👍

**Pull request merged**
